src/main/java/com/bobocode/hoverla/bring/context/BeanValidator.java

```diff
@@ -1,18 +1,113 @@
+                .collect(Collectors.groupingBy(BeanDefinition::name, Collectors.counting()))
+                .entrySet()
+                .stream()
+                .filter(v -> v.getValue() > 1)
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        if (!duplicateNames.isEmpty()) {
+            throw new BeanValidationException("Context contains beans with duplicate names: " + duplicateNames.keySet());
+        }
+    }
+
+    private void checkBeanName(BeanDefinition beanDefinition) {
+        log.trace("Checking name for bean definition: {} - {}", beanDefinition.name(), beanDefinition.type());
+        String name = beanDefinition.name();
+        if (beanDefinition.name() == null || name.isEmpty() || name.contains(" ")) {
+            throw new BeanValidationException(String.format("Bean name for %s class is null, empty or contains spaces",
+                    beanDefinition.type().getSimpleName()));
+        }
+    }
+
+    private void checkDependencies(BeanDefinition beanDefinition, List<BeanDefinition> beanDefinitions) {
+        log.trace("Checking dependencies for bean definition: {} - {}", beanDefinition.name(), beanDefinition.type());
+        Map<String, Class<?>> dependencies = beanDefinition.dependencies();
+        if (dependencies.isEmpty()) {
+            return;
+        }
+
+        for (var dependency : dependencies.entrySet()) {
+            log.trace("Checking dependency {} - {}", dependency.getKey(), dependency.getValue());
+            List<BeanDefinition> beansByName = findBeansByName(dependency.getKey(), beanDefinitions);
+            if (beansByName.size() > 1) {
+                throw new BeanValidationException("Context contains beans with duplicate names: " + dependency.getKey());
+            }
+            if (beansByName.size() == 1) {
+                checkTypeMatching(dependency.getValue(), beansByName.get(0));
+                checkCircularDependency(beanDefinition, beansByName.get(0));
+            } else {
+                checkBeansByType(dependency.getValue(), beanDefinitions);
+            }
+        }
+    }
+
+    private List<BeanDefinition> findBeansByName(String source, List<BeanDefinition> allBeanDefinitions) {
+        return allBeanDefinitions.stream()
+                .filter(b -> source.equals(b.name()))
+                .toList();
+    }
+
+    private void checkTypeMatching(Class<?> requiredType, BeanDefinition foundBeanDefinition) {
+        Class<?> foundType = foundBeanDefinition.type();
+        if (!foundType.equals(requiredType)) {
+            String exceptionMessage = "Types are not coincide for required type - %s and founded - %s";
+            throw new BeanValidationException(String.format(exceptionMessage, requiredType, foundType));
+        }
+    }
+
+    private void checkCircularDependency(BeanDefinition source, BeanDefinition found) {
+        if (found.dependencies().containsKey(source.name())) {
```

Comment:
don't we have to check inner dependencies of `found.dependencies()` recursively? so we won't have a situation with dependencies such as: A ->_(depends on)_ B, B -> C, C -> A. Don't we have a circular dependency here?

Reply:
Good point. I will cover this case. Thanks.